File: crates/vil_sidecar/sdk/vil_sidecar_sdk.php
#!/usr/bin/env php
<?php
/**
 * VIL Sidecar SDK — PHP
 * Connect to VIL host via UDS, exchange data via SHM, handle Invoke/Result.
 *
 * Usage:
 *   require_once 'vil_sidecar_sdk.php';
 *   $app = new VilSidecarApp('my-scorer');
 *   $app->handler('predict', function($data) { return ['score' => 0.95]; });
 *   $app->run();
 */
class VilSidecarApp {
    private string $name;
    private string $version;
    private array $handlers = [];
    private $conn;
    private $shmId = null;
    private int $shmSize = 0;

    public function __construct(string $name, string $version = '1.0.0') {
        $this->name = $name;
        $this->version = $version;
    }

    public function handler(string $method, callable $fn): self {
        $this->handlers[$method] = $fn;
        return $this;
    }

    public function run(): void {
        $socketPath = getenv('VIL_SIDECAR_SOCKET') ?: "/tmp/vil_sidecar_{$this->name}.sock";
        $this->conn = socket_create(AF_UNIX, SOCK_STREAM, 0);
        if (!socket_connect($this->conn, $socketPath)) {
            fwrite(STDERR, "[VIL Sidecar] Failed to connect to $socketPath\n");
            return;
        }

        // Send Handshake
        $this->send([
            'type' => 'Handshake', 'name' => $this->name, 'version' => $this->version,
            'methods' => array_keys($this->handlers), 'capabilities' => [], 'auth_token' => null,
        ]);

        // Receive HandshakeAck
        $ack = $this->recv();
        if (($ack['type'] ?? '') !== 'HandshakeAck' || !($ack['accepted'] ?? false)) {
            fwrite(STDERR, "[VIL Sidecar] Handshake rejected: " . ($ack['reject_reason'] ?? 'unknown') . "\n");
            return;
        }

        // Setup SHM
        $shmPath = $ack['shm_path'] ?? '';
        $this->shmSize = (int)($ack['shm_size'] ?? 0);
        if ($shmPath && file_exists($shmPath)) {
            $fd = fopen($shmPath, 'r+b');
            if ($fd) $this->shmId = $fd;
        }

        // Main loop
        while (true) {
            try {
                $msg = $this->recv();
            } catch (\Throwable $e) {
                break;
            }

            switch ($msg['type'] ?? '') {
                case 'Invoke': $this->handleInvoke($msg); break;
                case 'Health':
                    $this->send(['type' => 'HealthOk', 'in_flight' => 0,
                                 'total_processed' => 0, 'total_errors' => 0, 'uptime_secs' => 0]);
                    break;
                case 'Drain': $this->send(['type' => 'Drained']); break;
                case 'Shutdown': break 2;
            }
        }

        $this->cleanup();
    }

    private function handleInvoke(array $msg): void {
        $desc = $msg['descriptor'];
        $method = $msg['method'];
        $requestId = $desc['request_id'];

        // Read from SHM
        $inputData = [];
        if ($this->shmId && $desc['len'] > 0) {
            fseek($this->shmId, $desc['offset']);
            $raw = fread($this->shmId, $desc['len']);
            $inputData = json_decode($raw, true) ?: [];
        }

        $handler = $this->handlers[$method] ?? null;
        if (!$handler) {
            $this->sendResult($requestId, 'MethodNotFound', 0, 0, "no handler for '$method'");
            return;
        }

        try {
            $result = $handler($inputData);
            $resultBytes = json_encode($result);
            $respOffset = 1024 * 1024;
            if ($this->shmId) {
                fseek($this->shmId, $respOffset);
                fwrite($this->shmId, $resultBytes);
            }
            $this->sendResult($requestId, 'Ok', $respOffset, strlen($resultBytes));
        } catch (\Throwable $e) {
            $this->sendResult($requestId, 'Error', 0, 0, $e->getMessage());
        }
    }

    private function sendResult(int $requestId, string $status, int $offset = 0, int $len = 0, ?string $error = null): void {
        $this->send([
            'type' => 'Result', 'request_id' => $requestId, 'status' => $status,
            'descriptor' => $status === 'Ok' ? [
                'request_id' => $requestId, 'region_id' => 0, '_pad0' => 0,
                'offset' => $offset, 'len' => $len, 'method_hash' => 0, 'timeout_ms' => 0, 'flags' => 0,
            ] : null,
            'error' => $error,
        ]);
    }

    private function send(array $msg): void {
        $data = json_encode($msg);
        $header = pack('V', strlen($data));
        socket_write($this->conn, $header . $data);
    }

    private function recv(): array {
        $header = $this->recvExact(4);
        $len = unpack('V', $header)[1];
        $data = $this->recvExact($len);
        return json_decode($data, true) ?: [];
    }

    private function recvExact(int $n): string {
        $buf = '';
        while (strlen($buf) < $n) {
            $chunk = socket_read($this->conn, $n - strlen($buf));
            if ($chunk === false || $chunk === '') throw new \RuntimeException('connection closed');
            $buf .= $chunk;
        }
        return $buf;
    }

    private function cleanup(): void {
        if ($this->shmId) fclose($this->shmId);
        if ($this->conn) socket_close($this->conn);
    }
}

File: crates/vil_sidecar/sdk/vil_sidecar_sdk.py
#!/usr/bin/env python3
"""
VIL Sidecar SDK — Python
Connect to VIL host via UDS, exchange data via SHM, handle Invoke/Result protocol.

Usage:
    from vil_sidecar_sdk import SidecarApp
    app = SidecarApp("my-scorer")

    @app.handler("predict")
    def predict(data: dict) -> dict:
        return {"score": 0.95}

    app.run()
"""
import json, struct, socket, mmap, os, sys, time, traceback

class SidecarApp:
    def __init__(self, name: str, version: str = "1.0.0"):
        self.name = name
        self.version = version
        self.handlers = {}
        self.conn = None
        self.shm_fd = None
        self.shm_mm = None
        self.shm_size = 0

    def handler(self, method: str):
        """Decorator to register a handler for a method name."""
        def decorator(fn):
            self.handlers[method] = fn
            return fn
        return decorator

    def run(self):
        """Connect to VIL host, perform handshake, serve requests."""
        socket_path = os.environ.get("VIL_SIDECAR_SOCKET",
                                      f"/tmp/vil_sidecar_{self.name}.sock")
        self.conn = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        self.conn.connect(socket_path)

        # Send Handshake
        handshake = {
            "type": "Handshake",
            "name": self.name,
            "version": self.version,
            "methods": list(self.handlers.keys()),
            "capabilities": [],
            "auth_token": None,
        }
        self._send(handshake)

        # Receive HandshakeAck
        ack = self._recv()
        if ack.get("type") != "HandshakeAck" or not ack.get("accepted"):
            print(f"[VIL Sidecar] Handshake rejected: {ack.get('reject_reason')}", file=sys.stderr)
            return

        # Setup SHM
        shm_path = ack["shm_path"]
        self.shm_size = ack["shm_size"]
        self.shm_fd = os.open(shm_path, os.O_RDWR)
        self.shm_mm = mmap.mmap(self.shm_fd, self.shm_size)

        # Main loop
        while True:
            try:
                msg = self._recv()
            except (ConnectionError, struct.error):
                break

            msg_type = msg.get("type")
            if msg_type == "Invoke":
                self._handle_invoke(msg)
            elif msg_type == "Health":
                self._send({"type": "HealthOk", "in_flight": 0,
                            "total_processed": 0, "total_errors": 0,
                            "uptime_secs": 0})
            elif msg_type == "Drain":
                self._send({"type": "Drained"})
            elif msg_type == "Shutdown":
                break

        self._cleanup()

    def _handle_invoke(self, msg):
        desc = msg["descriptor"]
        method = msg["method"]
        request_id = desc["request_id"]
        offset = desc["offset"]
        length = desc["len"]

        # Read request data from SHM
        self.shm_mm.seek(offset)
        raw = self.shm_mm.read(length)
        try:
            input_data = json.loads(raw)
        except json.JSONDecodeError:
            input_data = raw.decode("utf-8", errors="replace")

        # Dispatch to handler
        handler = self.handlers.get(method)
        if handler is None:
            self._send_result(request_id, "MethodNotFound", error=f"no handler for '{method}'")
            return

        try:
            result = handler(input_data)
            result_bytes = json.dumps(result).encode("utf-8")

            # Write response to SHM (after request region)
            resp_offset = 1024 * 1024  # 1MB offset for responses
            self.shm_mm.seek(resp_offset)
            self.shm_mm.write(result_bytes)

            self._send_result(request_id, "Ok", resp_offset, len(result_bytes))
        except Exception as e:
            self._send_result(request_id, "Error", error=str(e))

    def _send_result(self, request_id, status, offset=0, length=0, error=None):
        msg = {
            "type": "Result",
            "request_id": request_id,
            "status": status,
            "descriptor": {"request_id": request_id, "region_id": 0, "_pad0": 0,
                           "offset": offset, "len": length,
                           "method_hash": 0, "timeout_ms": 0, "flags": 0} if status == "Ok" else None,
            "error": error,
        }
        self._send(msg)

    def _send(self, msg):
        data = json.dumps(msg).encode("utf-8")
        self.conn.sendall(struct.pack("<I", len(data)) + data)

    def _recv(self):
        raw_len = self._recv_exact(4)
        length = struct.unpack("<I", raw_len)[0]
        raw_data = self._recv_exact(length)
        return json.loads(raw_data)

    def _recv_exact(self, n):
        buf = b""
        while len(buf) < n:
            chunk = self.conn.recv(n - len(buf))
            if not chunk:
                raise ConnectionError("connection closed")
            buf += chunk
        return buf

    def _cleanup(self):
        if self.shm_mm:
            self.shm_mm.close()
        if self.shm_fd is not None:
            os.close(self.shm_fd)
        if self.conn:
            self.conn.close()
